packages/orm/src/EntityManager.ts

```diff
@@ -2006,79 +2006,124 @@ export class EntityManager<C = unknown, Entity extends EntityW = EntityW, TX ext
     }
-    // copy the context so that it's distinct between the two ems and so we can update any references from the old em
+    // copy the context so that it's distinct between the two ems, and so we can update any references from the old em
     // to the new one later
-    const ctx: C = { ...oldEm.ctx };
-    const newEm = new EntityManager<C, Entity, TX>(ctx, { em: oldEm });
+    const ctx = { ...oldEm.ctx } as Record<string, any>;
+    const newEm = new EntityManager<C, Entity, TX>(ctx as C, { em: oldEm });
+    if (allowPendingChanges) newEm.mode = "in-memory-writes";
     // hydrate (ie, create) each entity from the oldEm in the newEm
     oldEm.#entitiesByTag.forEach((entities, tag) => {
+      const [unpersistedEntities, persistedEntities] = allowPendingChanges
+        ? partition(entities, (e) => e.isNewEntity)
+        : [[], entities];
       // tags are always defined by the base constructor, so we use that as hydrate expects the base class as its argument
       const baseCstr = getConstructorFromTag(tag);
       // We copy the raw row data for each entity to avoid any potential conflict. Then we ensure the correct concrete
       // __class key is set for hydrate to use, if necessary (ie, cti), to actually instantiate the entity.
-      const rows = entities.map((e) => {
-        const instanceData = getInstanceData(e);
-        const meta = getMetadata(e);
-        const row = { ...instanceData.row };
-        if (!("id" in row)) {
-          // If we didn't get an id in our row data, then this entity was created by oldEm and flushed, so we have no
-          // row data and need to construct it manually
-          Object.values(meta.allFields).forEach((field) =>
-            field.serde?.columns?.forEach(
-              (column) => (row[column.columnName] = column.mapToDb(getField(e, field.fieldName))),
-            ),
-          );
-        }
-        if (meta.inheritanceType === "cti") row["__class"] = e.constructor.name;
-        return row;
-      });
+      const rows = persistedEntities.map((e) => createRowFromEntityData(e));
       newEm.hydrate(baseCstr, rows);
+      // Create blank entities in newEm for each unpersisted entity from oldEm.  They will be populated later in the
+      // allowPendingChanges step as they should not be present otherwise.
+      unpersistedEntities.forEach((e) => newEm.create(e.constructor as EntityConstructor<Entity>, {} as any));
     });
+
+    function findEntity(oldEntity: Entity): Entity {
+      if (oldEntity.isNewEntity) {
+        const index = parseInt(oldEntity.toString().split("#").pop()!) - 1;
+        return newEm.getEntities(oldEntity.constructor as EntityConstructor<Entity>)[index];
+      } else {
+        return newEm.getEntity(oldEntity.idTagged)!;
+      }
+    }
+
+    // If we are allowing pending changes, then we need to copy any changes across to newEm
+    if (allowPendingChanges) {
+      oldEm.entities.forEach((oldEntity) => {
```

Comment:
Is it worth filtering by `isDirtyEntity` here?

Reply:
not sure how much of a performance boost that'd be, but suppose it can't hurt